### PR TITLE
Add sstorage p2p protocol -- part 2

### PR DESCRIPTION
**Pull request by ping-ke**
doc about the sstorage p2p protocol, https://docs.google.com/document/d/1rfuGeCBU1Qj52gPOSyhefgb8hqLooTNPGiAeZbc85u4/edit;
and the part 1 code is https://github.com/QuarkChain/go-ethereum/pull/100.
Part 2 only contains the p2p sstorage protocol usage code.

- eth/
   - peer/peerset/handler/handler_sstorage
   - add sstorage peer to peerset
   - add sstorage handler to eth/handler
- eth/downloader/
   - add sstorage sync
- eth/ethconfig/
   - add sstorage config

### CODE REVIEW

File: eth/handler_sstorage.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/eth/protocols/sstorage"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// sstorageHandler implements the sstorage.Backend interface to handle the various network
+// packets that are sent as replies or broadcasts.
+type sstorageHandler handler
```
Comment (blockchaindevsh) [inline eth/handler_sstorage.go:27]: This `sstorageHandler` is not used yet?

File: eth/downloader/downloader.go [hunk 2]
```diff
@@ -322,6 +332,9 @@ func (d *Downloader) UnregisterPeer(id string) error {
 // adding various sanity checks as well as wrapping it with various log entries.
 func (d *Downloader) Synchronise(id string, head common.Hash, td *big.Int, mode SyncMode) error {
 	err := d.synchronise(id, head, td, mode)
+	if err == nil {
+		d.sstorStart <- struct{}{}
```
Comment (zhiqiangxu) [inline eth/downloader/downloader.go:336]: The design doc says `sstorage protocol starts after synced to the latest block, but here it seems not the case?

Comment (blockchaindevsh) [inline eth/downloader/downloader.go:336]: The design doc says `sstorage` protocol is supposed to be run after the node is synchronized to the latest block. But here it seems not the case?

File: eth/downloader/downloader.go [hunk 3]
```diff
@@ -1615,3 +1628,16 @@ func (d *Downloader) DeliverSnapPacket(peer *snap.Peer, packet snap.Packet) erro
 		return fmt.Errorf("unexpected snap packet type: %T", packet)
 	}
 }
+
+// DeliverSstoragePacket is invoked from a peer's message handler when it transmits a
+// data packet for the local node to consume.
+func (d *Downloader) DeliverSstoragePacket(peer *sstorage.Peer, packet sstorage.Packet) error {
+	switch packet := packet.(type) {
+	case *sstorage.KVsPacket:
+		fmt.Println(packet)
```
Comment (blockchaindevsh) [inline eth/downloader/downloader.go:1637]: Don't commit debug code.

File: eth/handler.go [hunk 4]
```diff
@@ -500,6 +533,7 @@ func (h *handler) unregisterPeer(id string) {
 
 func (h *handler) Start(maxPeers int) {
 	h.maxPeers = maxPeers
+	h.minPeersPerSstorageShard = 15 // todo update from config
```
Comment (blockchaindevsh) [inline eth/handler.go:536]: this parameter is not used yet?

File: eth/downloader/statesync.go [hunk 5]
```diff
@@ -74,6 +74,19 @@ func (d *Downloader) runStateSync(s *stateSync) *stateSync {
 	}
 }
 
+// sstorageFetcher manages the active sstorage sync and accepts requests
+// on its behalf.
+func (d *Downloader) sstorageFetcher() {
+	for {
+		select {
+		case <-d.sstorStart:
+			go d.SstorageSyncer.Sync(d.cancelCh)
```
Comment (blockchaindevsh) [inline eth/downloader/statesync.go:83]: Directly using this `d.cancelCh`  will have racing issue, preferring the snap sync style.

File: eth/handler.go [hunk 6]
```diff
@@ -316,16 +324,19 @@ func (h *handler) runEthPeer(peer *eth.Peer, handler eth.Handler) error {
 			}
 		}
 	}
+	if sstorExt != nil && sstorExt.Shards() == nil {
+		sstorExt.RequestShardList(sstor.Shards())
```
Comment (blockchaindevsh) [inline eth/handler.go:328]: Here `RequestShardList` is async, when `needThisPeer` is called, `sstorExt.Shards()` is still nil.

File: eth/peerset.go [hunk 7]
```diff
@@ -257,3 +375,23 @@ func (ps *peerSet) close() {
 	}
 	ps.closed = true
 }
+
+func (ps *peerSet) needThisPeer(peer *sstorage.Peer) bool {
```
Comment (qizhou) [inline eth/peerset.go:379]: Should we need ps lock here?

File: eth/handler.go [hunk 8]
```diff
@@ -459,6 +474,25 @@ func (h *handler) runSnapExtension(peer *snap.Peer, handler snap.Handler) error
...
 
+// runSstorageExtension registers a `sstorage` peer into the joint eth/sstorage peerset and
+// starts handling inbound messages. As `sstorage` is only a satellite protocol to
+// `eth`, all subsystem registrations and lifecycle management will be done by
+// the main `eth` handler to prevent strange races.
+func (h *handler) runSstorageExtension(peer *sstorage.Peer, handler sstorage.Handler) error {
+	h.peerWG.Add(1)
+	defer h.peerWG.Done()
+
+	if err := peer.Handshake(); err != nil {
```
Comment (qizhou) [inline eth/handler.go:485]: I am not sure this may create a race as the main eth protocol is also performing handshake at the same time.

Comment (ping-ke) [reply eth/handler.go:485]: main eth protocol will wait for sstorage handshake done.

Comment (qizhou) [reply eth/handler.go:485]: I would suggest adding a few comments to clarify the sequence of the handshake.

File: eth/protocols/sstorage/peer.go [hunk 9]
```diff
@@ -63,7 +63,10 @@ func (p *Peer) Shards() map[common.Address][]uint64 {
 
 // SetShards this should only be set when doing handshake.
 func (p *Peer) SetShards(shards map[common.Address][]uint64) {
-	p.shards = shards
+	// shards can only be set once.
+	if p.shards != nil {
```
Comment (qizhou) [inline eth/protocols/sstorage/peer.go:67]: I don't like hiding the issue - it should return an error and the caller should log an error for this.

**Review by qizhou — APPROVED** [review]

**Pull request merged**